app/validators/account_reset/granted_token_validator.rb

```diff
@@ -45,5 +45,13 @@ def account_reset_request
     def user
       account_reset_request&.user || AnonymousUser.new
     end
+
+    def requesting_issuer
+      account_reset_request&.requesting_issuer
+    end
+
+    def sp
+      @sp ||= ServiceProvider.find_by(issuer: requesting_issuer)
+    end
```

Comment:
1. We should add an association on the AccountResetRequest model for this, maybe call it `requesting_service_provider`
2. We want to no-op/skip the query if it's nil

on the model it might look like:

```ruby
belongs_to :requesting_service_provider,
           class_name: 'ServiceProvider',
           foreign_key: 'requesting_issuer',
           primary_key: 'issuer'

```

then this becomes

```suggestion
    def sp
      return @sp if defined?(@sp)
      @sp = account_reset_request.requesting_service_provider
    end
```

Reply:
Nice! thank you